prov/gni/src/gnix_msg.c

```diff
@@ -0,0 +1,297 @@
+	atomic_dec(&req->vc->outstanding_tx_reqs);
+	_gnix_vc_push_tx_reqs(req->vc);
+
+	_gnix_fr_free(ep, req);
+
+	return ret;
+
+}
+
+static int __comp_eager_msg_w_data_ack(void *data)
+{
+	return -FI_ENOSYS;
+}
+
+static int __comp_eager_msg_data_at_src(void *data)
+{
+	return -FI_ENOSYS;
+}
+
+static int __comp_eager_msg_data_at_src_ack(void *data)
+{
+	return -FI_ENOSYS;
+}
+
+static int __comp_rndzv_msg_rts(void *data)
+{
+	return -FI_ENOSYS;
+}
+
+static int __comp_rndzv_msg_rtr(void *data)
+{
+	return -FI_ENOSYS;
+}
+
+static int __comp_rndzv_msg_cookie(void *data)
+{
+	return -FI_ENOSYS;
+}
+
+static int __comp_rndzv_msg_send_done(void *data)
+{
+	return -FI_ENOSYS;
+}
+
+static int __comp_rndzv_msg_recv_done(void *data)
+{
+	return -FI_ENOSYS;
+}
+
+smsg_completer_fn_t gnix_ep_smsg_completers[] = {
+	[GNIX_SMSG_T_EGR_W_DATA] = __comp_eager_msg_w_data,
+	[GNIX_SMSG_T_EGR_W_DATA_ACK] = __comp_eager_msg_w_data_ack,
+	[GNIX_SMSG_T_EGR_GET] = __comp_eager_msg_data_at_src,
+	[GNIX_SMSG_T_EGR_GET_ACK] = __comp_eager_msg_data_at_src_ack,
+	[GNIX_SMSG_T_RNDZV_RTS] = __comp_rndzv_msg_rts,
+	[GNIX_SMSG_T_RNDZV_RTR] = __comp_rndzv_msg_rtr,
+	[GNIX_SMSG_T_RNDZV_COOKIE] = __comp_rndzv_msg_cookie,
+	[GNIX_SMSG_T_RNDZV_SDONE] = __comp_rndzv_msg_send_done,
+	[GNIX_SMSG_T_RNDZV_RDONE] = __comp_rndzv_msg_recv_done
+};
+
+static int _gnix_send_req(void *data)
+{
+	struct gnix_fab_req *req = (struct gnix_fab_req *)data;
+	struct gnix_nic *nic;
+	struct gnix_fid_ep *ep;
+	struct gnix_tx_descriptor *tdesc;
+	gni_return_t status;
+	int rc;
+	int rendevous = !!(req->flags & GNIX_MSG_RENDEVOUS);
+	int len;
+
+	ep = req->gnix_ep;
+	assert(ep != NULL);
+
+	nic = ep->nic;
+	assert(nic != NULL);
+
+	rc = _gnix_nic_tx_alloc(nic, &tdesc);
+	if (rc != FI_SUCCESS) {
+		GNIX_INFO(FI_LOG_EP_DATA, "_gnix_nic_tx_alloc() failed: %d\n",
+			  rc);
+		return -FI_EAGAIN;
+	}
+	assert(rc == FI_SUCCESS);
+
+	tdesc->desc.req = req;
+	tdesc->desc.ep = ep;
+	tdesc->desc.completer_fn =
+			gnix_ep_smsg_completers[GNIX_SMSG_T_EGR_W_DATA];
+
+	if (rendevous) {
+		/* Fill SMSG header with local buffer info. */
+
+		len = 0;
+		assert(0);
+	} else {
+		tdesc->desc.smsg_desc.hdr.len = req->len;
+		tdesc->desc.smsg_desc.hdr.flags = 0;
+		tdesc->desc.smsg_desc.buf = (void *)req->loc_addr;
+		len = req->len;
+	}
+
+	fastlock_acquire(&nic->lock);
+
+	status = GNI_SmsgSendWTag(req->vc->gni_ep,
+			&tdesc->desc.smsg_desc.hdr,
+			sizeof(struct gnix_smsg_hdr),
+			(void *)req->loc_addr, len,
+			tdesc->desc.id,
+			GNIX_SMSG_T_EGR_W_DATA);
+
+	fastlock_release(&nic->lock);
+
+	if (status == GNI_RC_NOT_DONE) {
+		_gnix_nic_tx_free(nic, tdesc);
+		GNIX_INFO(FI_LOG_EP_DATA,
+			  "GNI_SmsgSendWTag returned %s\n",
+			  gni_err_str[status]);
+	} else if (status != GNI_RC_SUCCESS) {
+		_gnix_nic_tx_free(nic, tdesc);
+		GNIX_WARN(FI_LOG_EP_DATA,
+			  "GNI_SmsgSendWTag returned %s\n",
+			  gni_err_str[status]);
+	}
+
+	return gnixu_to_fi_errno(status);
+}
+
+#define GNIX_MSG_RENDEVOUS_THRESH (16*1024)
```

Comment:
FWIW, the threshold will eventually become tunable with an `open_ops` function (#242) and this will become the default threshold value.